net/instaweb/http/redirect_following_url_async_fetcher.cc

```diff
@@ -0,0 +1,294 @@
+        redirect_following_fetcher_(redirect_following_fetcher),
+        recieved_redirect_status_code_(false),
+        urls_seen_(new GoogleUrlSet()),
+        url_(url),
+        gurl_(url),
+        base_fetch_(base_fetch),
+        context_url_(context_url),
+        message_handler_(message_handler) {
+    GoogleString sanitized = GoogleUrl::Sanitize(url);
+    urls_seen_->insert(sanitized);
+  }
+
+  RedirectFollowingFetch(
+      RedirectFollowingUrlAsyncFetcher* redirect_following_fetcher,
+      AsyncFetch* base_fetch, const GoogleString& url,
+      const GoogleString& context_url, GoogleUrlSet* redirects_followed_earlier,
+      MessageHandler* message_handler)
+      : SharedAsyncFetch(base_fetch),
+        redirect_following_fetcher_(redirect_following_fetcher),
+        recieved_redirect_status_code_(false),
+        urls_seen_(redirects_followed_earlier),
+        url_(url),
+        gurl_(url),
+        base_fetch_(base_fetch),
+        context_url_(context_url),
+        message_handler_(message_handler) {}
+
+  bool Validate() {
+    if (!gurl_.IsWebValid()) {
+      response_headers()->set_status_code(HttpStatus::kBadRequest);
+      return false;
+    }
+    return true;
+  }
+
+ protected:
+  virtual bool HandleFlush(MessageHandler* message_handler) {
+    if (!recieved_redirect_status_code_) {
+      return SharedAsyncFetch::HandleFlush(message_handler);
+    }
+    return true;
+  }
+
+  virtual void HandleHeadersComplete() {
+    GoogleString h = response_headers()->ToString();
+    // Currently we support permanent and temporary redirects.
+    recieved_redirect_status_code_ =
+        response_headers()->status_code() == HttpStatus::kMovedPermanently ||
+        response_headers()->status_code() == HttpStatus::kFound;
+
+    if (!recieved_redirect_status_code_) {
+      SharedAsyncFetch::HandleHeadersComplete();
+    }
+  }
+
+  virtual bool HandleWrite(const StringPiece& content,
+                           MessageHandler* handler) {
+    if (!recieved_redirect_status_code_) {
+      return SharedAsyncFetch::HandleWrite(content, handler);
+    }
+    return true;
+  }
+
+  virtual void HandleDone(bool success) {
+    DCHECK(gurl_.IsWebValid() || !success);
+    const RewriteOptions* options =
+        redirect_following_fetcher_->rewrite_options();
+
+    if (!recieved_redirect_status_code_) {
+      SharedAsyncFetch::HandleDone(success);
+      delete this;
+      return;
+    }
+
+    bool my_success = success;
+    int redirects_followed = urls_seen_->size();
+
+    if (my_success) {
+      // TODO(oschaaf): does this have the correct arguments?
+      bool cacheable = response_headers()->IsProxyCacheable(
+          base_fetch_->request_headers()->GetProperties(),
+          ResponseHeaders::GetVaryOption(options->respect_vary()),
+          ResponseHeaders::kHasValidator);
+      if (!cacheable) {
+        LOG(WARNING) << "Not following uncacheable redirect: " << url_;
+        my_success = false;
+      }
```

Comment:
@morlovich I have worked on combining cacheability. 
I need to add tests for this, and address the other comments.

As for 302/307 responses being cacheable: I think they are but only when the headers explicitly say so. That is the behaviour this pr implements.

Some references

https://support.cloudflare.com/hc/en-us/articles/200168326-Are-301-and-302-redirects-cached-by-CloudFlare-
https://community.akamai.com/thread/1284
https://aprescott.com/posts/http-redirects